src/librustc_typeck/check/_match.rs

```diff
@@ -27,18 +27,96 @@ use syntax::ptr::P;
 use syntax_pos::Span;
 
 impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
-    pub fn check_pat(&self, pat: &'gcx hir::Pat, expected: Ty<'tcx>) {
-        self.check_pat_arg(pat, expected, false);
+    pub fn check_pat(&self, pat: &'gcx hir::Pat, expected: Ty<'tcx>, def_bm: ty::BindingMode) {
+        self.check_pat_arg(pat, expected, def_bm, false);
     }
 
     /// The `is_arg` argument indicates whether this pattern is the
     /// *outermost* pattern in an argument (e.g., in `fn foo(&x:
     /// &u32)`, it is true for the `&x` pattern but not `x`). This is
     /// used to tailor error reporting.
-    pub fn check_pat_arg(&self, pat: &'gcx hir::Pat, expected: Ty<'tcx>, is_arg: bool) {
+    pub fn check_pat_arg(&self, pat: &'gcx hir::Pat, mut expected: Ty<'tcx>,
+                         mut def_bm: ty::BindingMode, is_arg: bool) {
         let tcx = self.tcx;
 
-        debug!("check_pat(pat={:?},expected={:?},is_arg={})", pat, expected, is_arg);
+        debug!("check_pat(pat={:?},expected={:?},def_bm={:?},is_arg={})",
+            pat, expected, def_bm, is_arg);
+
+        let is_non_ref_pat = match pat.node {
+            PatKind::Struct(..) |
+            PatKind::TupleStruct(..) |
+            PatKind::Tuple(..) |
+            PatKind::Box(_) |
+            // PatKind::Lit(_) | // FIXME(tschottdorf): causes lots of errors
+            PatKind::Range(..) |
+            PatKind::Slice(..) => true,
+            PatKind::Path(ref qpath) => {
+                // FIXME(tschottdorf): is it OK to call this here?
+                let (def, _, _) = self.resolve_ty_and_def_ufcs(qpath, pat.id, pat.span);
```

Comment:
Hmm. It's not great to call this here, in that it can report an error, which might then get reported again later on (by `check_pat_path`). Not sure how best to fix this though!

Reply:
Hmm, we might be able to just modify `resolve_ty_and_def_ufcs` to cache the result. It already writes the final result into a table:

```rust
// Write back the new resolution.
let hir_id = self.tcx.hir.node_to_hir_id(node_id);
self.tables.borrow_mut().type_dependent_defs_mut().insert(hir_id, def);
```

seems like we can check that cache in the case that we get `TypeRelative`. 

(cc @eddyb -- this seem ok to you?)